### PR TITLE
Internal code refactoring

### PR DESCRIPTION
**Pull request by Windvis**
- Use a Glimmer-like setup: outer-html semantics and lifecycle modifiers. This makes updating to Glimmer components a lot easier in the future.
- use containers instead of instances to refer to the breadcrumbs component element.
- Add methods to the service to manage containers instead of directly modifying the service state

Let me know what you think, this is just my vision of things. Service methods still need tests.

### CODE REVIEW

File: addon/components/breadcrumbs.js [hunk 1]
```diff
@@ -21,17 +21,26 @@ export default Component.extend({
    */
 
   layout,
-  tagName: 'ul',
+  tagName: '',
+  element: null,
```
Comment (bertdeblock) [inline addon/components/breadcrumbs.js]: Any reason for this?

File: addon/components/breadcrumbs-item.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import Component from '@ember/component'
 import { inject as service } from '@ember/service'
 import layout from '../templates/components/breadcrumbs-item'
+import { alias } from '@ember/object/computed'
```
Comment (bertdeblock) [inline addon/components/breadcrumbs-item.js]: `readOnly` would be more correct?

File: addon/components/breadcrumbs.js [hunk 3]
```diff
@@ -21,17 +21,26 @@ export default Component.extend({
...
+    })
+
+    this.breadcrumbsService.registerContainer(this.containerData)
   },
 
-  willDestroyElement () {
-    this.get('breadcrumbsService.instances').removeObject(this)
+  handleDestroy () {
+    this.breadcrumbsService.unregisterContainer(this.containerData)
+    this.set('containerData', null)
```
Comment (bertdeblock) [inline addon/components/breadcrumbs.js]: Wonder if this is necessary? Isn't the ref to `containerData` removed when the component gets destroyed?

File: addon/services/breadcrumbs.js [hunk 4]
```diff
@@ -15,6 +15,40 @@ export default Service.extend({
...
+    this.containers = []
+  },
+
+  /**
+   * Methods
+   */
+
+  registerContainer (containerData) {
+    assert(
+      'An `element` is required to register a breadcrumbs container',
```
Comment (bertdeblock) [inline addon/services/breadcrumbs.js]: `a breadcrumb container`?

File: addon/services/breadcrumbs.js [hunk 5]
```diff
@@ -15,6 +15,40 @@ export default Service.extend({
...
+    )
+
+    this.set('containers', [...this.containers, containerData])
+  },
+
+  unregisterContainer ({ element }) {
+    this.set(
+      'containers',
+      this.containers.filter(registeredContainer => {
+        return element === registeredContainer.element
```
Comment (bertdeblock) [inline addon/services/breadcrumbs.js]: `!==` no?

File: addon/components/breadcrumbs.js (hunk 1 above)
Comment (Windvis) [reply addon/components/breadcrumbs.js]: forgot to remove it. In the first refactor I still passed the component instance to the service, and BreadcrumbsItem required a .element prop.

File: addon/components/breadcrumbs-item.js (hunk 2 above)
Comment (Windvis) [reply addon/components/breadcrumbs-item.js]: true

File: addon/components/breadcrumbs.js (hunk 3 above)
Comment (Windvis) [reply addon/components/breadcrumbs.js]: No it probably isn't necessary

File: addon/services/breadcrumbs.js (hunk 5 above)
Comment (Windvis) [reply addon/services/breadcrumbs.js]: heh, that's what happens when you don't write tests :P

**Comment by Windvis** [comment]
Added some service tests, should be good to gooooo.

File: addon/services/breadcrumbs.js [hunk 6]
```diff
@@ -15,6 +15,42 @@ export default Service.extend({
...
+  unregisterContainer ({ element }) {
+    this.set(
+      'containers',
+      this.containers.filter(registeredContainer => {
+        return element !== registeredContainer.element
+      })
+    )
+  },
+
+  /* Private */
```
Comment (Windvis) [inline addon/services/breadcrumbs.js]: should we just start prefixing with `_` like most js devs?

Comment (bertdeblock) [reply addon/services/breadcrumbs.js]: Not sure, could do, slightly indifferent about it.

File: addon/services/breadcrumbs.js [hunk 7]
```diff
@@ -15,6 +15,42 @@ export default Service.extend({
...
+
+  isContainerRegistered ({ element }) {
+    let containerData = this.containers.find(registeredContainer => {
```
Comment (bertdeblock) [inline addon/services/breadcrumbs.js]: Could also use `some`? It already returns a Boolean => https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some. Not important though.

File: tests/unit/services/breadcrumbs-test.js [hunk 8]
```diff
@@ -4,9 +4,58 @@ import { setupTest } from 'ember-qunit'
...
+    assert.throws(function () {
+      service.registerContainer(containerData)
+    })
+  })
+
+  test('it throws an error if no element is provided', function (assert) {
+    let service = this.owner.lookup('service:breadcrumbs')
+
+    assert.throws(function () {
+      let nonValidContainerData = {}
```
Comment (bertdeblock) [inline tests/unit/services/breadcrumbs-test.js]: `inValidContainerData`?

**Comment by bertdeblock** [comment]
Done. Only 2 small remarks. Looks tight!

File: addon/services/breadcrumbs.js (hunk 7 above)
Comment (bertdeblock) [reply addon/services/breadcrumbs.js]: And maybe find is more clear in this case anyways.

**Comment by bertdeblock** [comment]
Can you clean up the commits as well (correct prefixes and stuffz)?

**Comment by Windvis** [comment]
> Can you clean up the commits as well (correct prefixes and stuffz)?

Yea, was waiting for feedback. And squash merging is also an option.

File: addon/services/breadcrumbs.js (hunk 6 above)
Comment (Windvis) [reply addon/services/breadcrumbs.js]: I hate it, but it does make it obvious which things should not be touched from the outside. I just which JS came up with native private scoping (without h4x).

File: addon/services/breadcrumbs.js (hunk 7 above)
Comment (Windvis) [reply addon/services/breadcrumbs.js]: Nah, some is better

File: addon/services/breadcrumbs.js (hunk 6 above)
Comment (bertdeblock) [reply addon/services/breadcrumbs.js]: Yup, underscoring is fine for me as well.

**Comment by Windvis** [comment]
> > Can you clean up the commits as well (correct prefixes and stuffz)?
> 
> Yea, was waiting for feedback. And squash merging is also an option.

Rewrote history, timestamps are messed up though :smile:

File: addon/services/breadcrumbs.js (hunk 7 above)
Comment (bertdeblock) [reply addon/services/breadcrumbs.js]: Forgot to add this one it seems?

Comment (Windvis) [reply addon/services/breadcrumbs.js]: Yeaaaa, rolled back when debugging the test issue, and forgot to reapply it seems.

Comment (Windvis) [reply addon/services/breadcrumbs.js]: Should be good good now

**Pull request merged**